wave_optimization/include/wave/optimization/ceres/ba.hpp

```diff
@@ -0,0 +1,108 @@
+#ifndef WAVE_OPTIMIZATION_CERES_BA_HPP
+#define WAVE_OPTIMIZATION_CERES_BA_HPP
+
+#include <typeinfo>
+
+#include <ceres/ceres.h>
+
+#include "wave/utils/utils.hpp"
+#include "wave/vision/utils.hpp"
+#include "wave/vision/dataset.hpp"
+
+namespace wave {
+
+/** Bundle Adjustment Residual
+ */
+struct BAResidual {
+ public:
+    double fx;
+    double fy;
+    double cx;
+    double cy;
+
+    double x;
+    double y;
+
+    BAResidual() : fx(0.0), fy(0.0), cx(0.0), cy(0.0), x(0.0), y(0.0) {}
+
+    BAResidual(Mat3 K, Vec2 x)
+        : fx(K(0, 0)),
+          fy(K(1, 1)),
+          cx(K(0, 2)),
+          cy(K(1, 2)),
+          x(x(0)),
+          y(x(1)) {}
+
+    /**
+       * Calculate Bundle Adjustment Residual
+       *
+     * @param q_GC Camera rotation parameterized with a quaternion (x, y, z, w)
+     * @param G_p_GC Camera translation (x, y, z)
+     * @param world_pt World point (x, y, z)
```

Comment:
We should merge in that [geometry notation markdown page](https://github.com/wavelab/libwave/blob/issue_33_geometry_notation/docs/ref/geometry/geometry_notation.md). That explains the "G_p_GC", etc., notation.

That page should specify that the world frame "G" is oriented NWU - e.g. if you're walking along, the third axis points up - but that doesn't matter for these particular calculations.

The distinction previously used in this file between NWU and EDN is not needed. The camera frame is related to the body frame by a regular old rotation (which can be expressed as a rotation marix or quaternion), not any special "nwu2edn".